File: Life/Domain/Game.cs
#region Usings

using System;
using System.Text;

#endregion

namespace Domain {
    [Serializable]
    public class Game {
        [NonSerialized]
        public bool[,] Cells;
        public int Size { get; private set; }
        public string Name { get; private set; }

        public Game(int size, string name) {
            Size = size;
            Name = name;
            Cells = new bool[size,size];
            for (var row = 0; row < Size; row++) {
                for (var column = 0; column < Size; column++) {
                    Cells[row, column] = false;
                }
            }
        }

        public Game(int size) : this(size, Guid.NewGuid().ToString()) {}

        public Cell this[int row, int column] {
            get { return new Cell(Cells[row, column]); }
        }

        public void GiveBirth(int row, int column) {
            Cells[row, column] = true;
        }

        public void Step() {
            var newGeneration = new bool[Size,Size];
            for (var row = 0; row < Size; row++) {
                for (var column = 0; column < Size; column++) {
                    newGeneration[row, column] = false;
                    if (this[row, column].IsAlive && NumberOfNeighbors(row, column) < 2) {
                        newGeneration[row, column] = false;
                    }
                    if (this[row, column].IsAlive && NumberOfNeighbors(row, column) == 2) {
                        newGeneration[row, column] = true;
                    }
                    if (this[row, column].IsAlive && NumberOfNeighbors(row, column) == 3) {
                        newGeneration[row, column] = true;
                    }
                    if (this[row, column].IsAlive && NumberOfNeighbors(row, column) == 4) {
                        newGeneration[row, column] = false;
                    }
                    if (this[row, column].IsDead && NumberOfNeighbors(row, column) == 3) {
                        newGeneration[row, column] = true;
                    }
                }
            }
            Cells = newGeneration;
        }

        public override string ToString() {
            var sb = new StringBuilder();
            for (var row = 0; row < Size; row++) {
                for (var column = 0; column < Size; column++) {
                    sb.Append(this[row, column].IsAlive ? 'x' : '.');
                }
                sb.AppendLine();
            }
            return sb.ToString().TrimEnd(Environment.NewLine.ToCharArray());
        }

        private int NumberOfNeighbors(int row, int column) {
            return
                Neighbor(row - 1, column - 1) +
                Neighbor(row - 1, column) +
                Neighbor(row - 1, column + 1) +
                Neighbor(row, column - 1) +
                Neighbor(row, column + 1) +
                Neighbor(row + 1, column - 1) +
                Neighbor(row + 1, column) +
                Neighbor(row + 1, column + 1);
        }

        private int Neighbor(int row, int column) {
            if (row < 0 || row >= Size || column < 0 || column >= Size) return 0;
            return Cells[row, column] ? 1 : 0;
        }
    }
}

File: life-java/src/main/java/ru/agiledays/life/server/domain/Game.java
package ru.agiledays.life.server.domain;

import java.util.UUID;

public class Game {
    // NB! This field should become private after refactoring to perfect tests
    public boolean[][] Cells;
    
    private final int size;
    private final String name;
    
    public Game(final int size, final String name) {
        this.size = size;
        this.name = name;
        
        Cells = createCells();
    }
    
    public Game(final int size) {
        this.size = size;
        this.name = UUID.randomUUID().toString();
        
        Cells = createCells();
    }
    
    public int size() {
        return size;
    }
    
    public String name() {
        return name;
    }
    
    public Cell getCell(final int row, final int column) {
        return new Cell(Cells[row][column]);
    }
    
    public void giveBirth(final int row, final int column) {
        Cells[row][column] = true;
    }
    
    public void step() {
        final boolean[][] newGeneration = createCells();
        
        for (int row = 0; row < size; row++) {
            for (int column = 0; column < size; column++) {
                newGeneration[row][column] = false;
                if (getCell(row, column).isAlive() && numberOfNeighbors(row, column) < 2) {
                    newGeneration[row][column] = false;
                }
                if (getCell(row, column).isAlive() && numberOfNeighbors(row, column) == 2) {
                    newGeneration[row][column] = true;
                }
                if (getCell(row, column).isAlive() && numberOfNeighbors(row, column) == 3) {
                    newGeneration[row][column] = true;
                }
                if (getCell(row, column).isAlive() && numberOfNeighbors(row, column) == 4) {
                    newGeneration[row][column] = false;
                }
                if (getCell(row, column).isDead() && numberOfNeighbors(row, column) == 3) {
                    newGeneration[row][column] = true;
                }
            }
        }
        
        Cells = newGeneration;
    }

    @Override
    public String toString() {
        final StringBuffer sb = new StringBuffer();
        for (int row = 0; row < size; row++) {
            if (row != 0) {
                sb.append(StringHelper.newLine);
            }
            for (int column = 0; column < size; column++) {
                sb.append(getCell(row, column).isAlive() ? 'x' : '.');
            }
        }
        return sb.toString();
    }

    private int numberOfNeighbors(final int row, final int column) {
        return
            neighbor(row - 1, column - 1) +
            neighbor(row - 1, column) +
            neighbor(row - 1, column + 1) +
            neighbor(row, column - 1) +
            neighbor(row, column + 1) +
            neighbor(row + 1, column - 1) +
            neighbor(row + 1, column) +
            neighbor(row + 1, column + 1);
    }

    private int neighbor(final int row, final int column) {
        if (row < 0 || row >= size || column < 0 || column >= size) {
            return 0;
        }
        return Cells[row][column] ? 1 : 0;
    }
    
    private boolean[][] createCells() {
        return new boolean[size][size];
    }
}
